### PR TITLE
Wrangler decimal improvement and cross column arithmetic operations

### PR DESCRIPTION
**Pull request by flakrimjusufi**
Wrangler - Decimal Transformation Improvement and Cross Column Operations
JIRA Ticket: https://cdap.atlassian.net/browse/PLUGIN-598

### CODE REVIEW

**Comment by CuriousVini** [comment]
Please add jira to the description

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    if (x == null || y == null) {
+      return null;
+    }
+    return x.multiply(y);
+  }
+
+  /**
+   Arithmetic operation - Divide a column by another and return the quotient.
+   */
+  public static Integer divideq(Integer x, Integer y) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java:147]: is there any specific reason for naming it as `divideq`

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java [hunk 2]
```diff
@@ -241,6 +278,8 @@ private static Object convertType(String col, String toType, Object object) thro
           return Bytes.toBytes((Integer) object);
         } else if (object instanceof Long) {
           return Bytes.toBytes((Long) object);
+        } else if (object instanceof BigDecimal) {
+          return Bytes.toBytes((BigDecimal) object);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:282]: Is this same as getting bigdecimal unscaled value and getting bytesarray out of it?

File: wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java [hunk 3]
```diff
@@ -141,24 +144,52 @@ public void testToDouble() throws Exception {
...
 
+  @Test
+  public void testToDecimal() throws Exception {
+    List<Row> rows = Collections.singletonList(
+      new Row("str_col", "10000").add("int_col", 10000).add("double_col", 10000.00d)
+        .add("short_col", new Short("10000")).add("long_col", 10000L).add("float_col", 10000.0f)
+        .add("bytes_col", new byte[] {0, 0, 0, 0, 39, 16})
+    );
+    String[] directives = new String[] {
+      "set-type str_col decimal", "set-type int_col Decimal", "set-type double_col DECIMAL",
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java:155]: should include all the types in the test

**Review by CuriousVini — COMMENTED** [review]
1. Please include jira to the description and PR to the jira. 
2. Please add documentation on how these functions can be used by users.
3. What is the current behavior when decimal type is used in set type directive?

**Comment by flakrimjusufi** [comment]
> 1. Please include jira to the description and PR to the jira.
> 2. Please add documentation on how these functions can be used by users.
> 3. What is the current behavior when decimal type is used in set type directive?

1. PR updated. 
2. PR updated. 
3. The current behavior when decimal type is used in set type directive is getting unsupported type error such as: 

`"Column '%s' is of unsupported type '%s'. Supported types are: " +
            "int, short, long, double, boolean, string, bytes"`

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java (hunk 1 above)
Comment (flakrimjusufi) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java:147]: This is how it was named in PRD. 

Please check it out:  

image

File: wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java (hunk 3 above)
Comment (flakrimjusufi) [reply wrangler-core/src/test/java/io/cdap/directives/column/SetTypeTest.java:155]: PR Updated.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java (hunk 2 above)
Comment (flakrimjusufi) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:282]: Actually, converting _Bytes_ to _BigDecimal_ and the other way round was not included in the PRD. 

image

We added this conversion unintentionally when we were doing some unit tests.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java (hunk 1 above)
Comment (CuriousVini) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java:147]: I see

File: wrangler-docs/directives/set-column.md [hunk 4]
```diff
@@ -59,3 +59,42 @@ would result in this record:
...
+  "hrlywage": "11.79",
+  "yrlywage": "21151.26"
+  "yrlyhours": "1794"
+  "yrlybonus": "25000.98"
+  "wageincrease": "2.50"
+}
+```
+Applying these directives:
+```
+set-column newhrlywage arithmetic:add(hrlywage,wageincrease)
```
Comment (CuriousVini) [inline wrangler-docs/directives/set-column.md]: should list all the functions available in the documentation

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java (hunk 2 above)
Comment (CuriousVini) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:282]: We should keep this. but is `Bytes.toBytes((BigDecimal) object)` same as  getting bigdecimal unscaled value and getting bytesarray out of it?

**Review by CuriousVini — COMMENTED** [review]
A few minor comments. Fix and squash

File: wrangler-docs/directives/set-column.md (hunk 4 above)
Comment (flakrimjusufi) [reply wrangler-docs/directives/set-column.md]: All the new functions are included in documentation in the last update.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java (hunk 2 above)
Comment (flakrimjusufi) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:282]: No, they are not the same. 

We are using this method to transform bytes to decimal: 

https://github.com/cdapio/cdap/blob/214a22b4d5d80baa7b5b4ac11d6328ff22864498/cdap-api-common/src/main/java/io/cdap/cdap/api/common/Bytes.java#L767-L772

**Review by CuriousVini — APPROVED** [review]

**Comment by flakrimjusufi** [comment]
@CuriousVini 

Fixed checkstyle issue and squashed the commits.

**Pull request merged**